src/aof.c

```diff
@@ -40,191 +40,603 @@
+ *       AOFs will become HISTORY. they will be cleaned regularly in cron. 
+ * INCR: There may be more than one (after multiple AOFRW failures), and they 
+ *       together represent all the incremental commands executed by redis after
+ *       the last AOFRW.
+ * 
+ * The following is a possible aof manifest file content:
+ * 
+ * fileName appendonly.2.rdb fileSeq 2 fileType b
+ * fileName appendonly.1.aof fileSeq 1 fileType h
+ * fileName appendonly.2.aof fileSeq 2 fileType h
+ * fileName appendonly.3.aof fileSeq 3 fileType h
+ * fileName appendonly.4.aof fileSeq 4 fileType i
+ * fileName appendonly.5.aof fileSeq 5 fileType i
  * ------------------------------------------------------------------------- */
 
-#define AOF_RW_BUF_BLOCK_SIZE (1024*1024*10)    /* 10 MB per block */
-
-typedef struct aofrwblock {
-    unsigned long used, free, pos;
-    /* Note that 'buf' must be the last field of aofrwblock struct, because
-     * memory allocator may give us more memory than our apply for reducing
-     * fragments, but we want to make full use of given memory, i.e. we may
-     * access the memory after 'buf'. To avoid make others fields corrupt,
-     * 'buf' must be the last one. */
-    char buf[AOF_RW_BUF_BLOCK_SIZE];
-} aofrwblock;
-
-/* This function free the old AOF rewrite buffer if needed, and initialize
- * a fresh new one. It tests for server.aof_rewrite_buf_blocks equal to NULL
- * so can be used for the first initialization as well. */
-void aofRewriteBufferReset(void) {
-    if (server.aof_rewrite_buf_blocks)
-        listRelease(server.aof_rewrite_buf_blocks);
-
-    server.aof_rewrite_buf_blocks = listCreate();
-    listSetFreeMethod(server.aof_rewrite_buf_blocks,zfree);
+/* Naming rules */
+#define BASE_AOF_SUFFIX          ".rdb"        
+#define INCR_AOF_SUFFIX          ".aof"        
+#define MANIFEST_NAME_SUFFIX     ".manifest"   
+#define MANIFEST_TEM_NAME_PREFIX "temp_"      
+
+/* AOF manifest key */
+#define AOF_MANIFEST_KEY_FILE_NAME   "fileName"
+#define AOF_MANIFEST_KEY_FILE_SEQ    "fileSeq"
+#define AOF_MANIFEST_KEY_FILE_TYPE   "fileType"
+
+/* Create an empty aofInfo. */
+aofInfo *aofInfoCreate(void) {
+    return zcalloc(sizeof(aofInfo));
+}
+
+/* Free the aofInfo structure (pointed to by ai)  and its embedded file_name. */
+void aofInfoFree(aofInfo *ai) {
+    serverAssert(ai != NULL);
+    if (ai->file_name) sdsfree(ai->file_name);
+    zfree(ai);
+}
+
+/* Deep copy an aofInfo. */
+aofInfo *aofInfoDup(aofInfo *orig) {
+    serverAssert(orig != NULL);
+    aofInfo *ai = aofInfoCreate();
+    ai->file_name = sdsdup(orig->file_name);
+    ai->file_seq = orig->file_seq;
+    ai->file_type = orig->file_type;
+    return ai;
+}
+
+/* Method to free AOF list elements. */
+void aofListFree(void *item) {
+    aofInfo *ai = (aofInfo *)item;
+    aofInfoFree(ai);
+}
+
+/* Method to duplicate AOF list elements. */
+void *aofListDup(void *item) {
+    return aofInfoDup(item);;
```

Comment:
```suggestion
    return aofInfoDup(item);
```

p.s. i think we can give up this thin wrapper, and just copy the dup code here (we don't use it elsewhere)

Reply:
Also used in `aofManifestDup` to dup base_aof_info